docs/dump_heapdump.md

```diff
@@ -24,25 +24,354 @@
+
+
+#### Class records
+
+The PHD class record encodes a class object and contains the following format:
+
+- The 1 byte tag, containing the value 6.
+
+- A byte containing the following flags:
+
+    | Bit number | Value or description                                                                      |
+    |------------|-------------------------------------------------------------------------------------------|
+    | 1 and 2    | Indicates whether the gap is byte, `short`, `int` or `long`                               |
+    | 3 and 4    | Indicates the size of each static reference  (0=`byte`, 1=`short`, 2=`int`, 3=`long`)     |
+    | 5          | Indicates if the object was hashed                                                        |
+
+- A byte, `short`, `int` or `long` containing the gap between the address of this class and the address of the preceding object (See the **Short object record** description).
+- An `int` containing the instance size.
+- If all objects are hashed, a `short` containing the hash code. Otherwise, an optional `int` containing the hash code if the hashed and moved bit is set in the record flag byte.
+- A `word` containing the address of the superclass.
+- A UTF string containing the name of this class.
+- An `int` containing the number of static references.
+- The array of static references (See the **Short object record** description).
+
+### Classic Heap Dump format
+
+Classic heap dumps are produced in ascii text on all platforms except z/OS, which are encoded in EBCDIC. The dump is divided into the following sections:
+
+#### Header record
+
+A single string containing information about the runtime environment, platform, and build levels, similar to the following example:
+
+```
+// Version: JRE 1.8.0 Linux amd64-64 (build 1.8.0_232-b09)
+```
+
+#### Object records
+
+A record of each object instance in the heap with the following format:
+
+```
+<object address, in hexadecimal> [<length in bytes of object instance, in decimal>] OBJ <object type>
+<heap reference, in hexadecimal> <heap reference, in hexadecimal> ...
+```
+
+The following object types (`object type`) might be shown:
+
+- class name (including package name)
+- class array type
+- primitive array type
+
+These types are abbreviated in the record. To determine the type, see the [Java VM Type Signature table](#java-vm-type-signatures).
+
+Any references found are also listed, excluding references to an object's class or NULL references.
+
+The following example shows an object instance (16 bytes in length) of type `java/lang/String`, with a reference to a char array:
+
+```
+0x00000000E0000AF0 [16] OBJ java/lang/String
+	0x00000000E0000B00
+```
+
+The object instance (length 32 bytes) of type char array, as referenced from the `java/lang/String`, is shown in the following example:
+
+```
+0x00000000E0000B00 [32] OBJ [C
+```
+
+The following example shows an object instance (24 bytes in length) of type array of `java/lang/String`:
```

Comment:
The length 24 doesn't make sense, since there are 32 bytes of data shown.

Reply:
I thought the length seemed to correlate with the `[xx]` value, although 24 is very odd:

```
0x00000000FFF07498 [24] OBJ [Ljava/lang/String;
	0x00000000E0005D78 0x00000000E0005D50 0x00000000E0005D28 0x00000000E0005D00
```

What does the square bracket indicate?

Reply:
Either change the 24's to 32's, or remove one of the four array entries.
The `[` in `[Ljava/lang/String;` means it's an array of String objects.

Reply:
Hmm, there is obviously something I'm not understanding about the format, as the following is taken from a classic heap dump I just created. `(24 bytes in length)` doesn't match having 4 array entries of 8 bytes. I think my confusion is that although they are printed as 8 byte values, since this is compressed refs they are really 4 byte values. 
```
0x00000000FFDF1F60 [24] OBJ [Ljava/lang/String;
	0x00000000DFFF5790 0x00000000DFFF5768 0x00000000DFFF5740 0x00000000DFFF5718 
```